### ISSUE TITLE
Symfony 2.8 explodes on "cold" app/console start when session storage is configured with PDO

### ISSUE THREAD

**Issue by sergeil** [symfony/symfony#18443]
If Symfony is configured to use PDO session storage and no cache is available yet (app/cache/[env]/*), then when trying to run app/console for the first time, it explodes by trying to connect to a database (if specified database connection credentials are not valid OR database is not reachable).  With Symfony 2.7 it works fine.

```
~/symfony28 $ sudo rm app/cache/* -rf && docker run -it --rm -v `pwd`:/mnt/tmp -w /mnt/tmp modera/php5-fpm:1.0 app/console -vvv
                                                                             
  [PDOException]                                                             
  SQLSTATE[HY000] [2003] Can't connect to MySQL server on '127.0.0.1' (111)  
                                                                             

Exception trace:
 () at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:1693
 PDO->__construct() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:1693
 appDevDebugProjectContainer->getPdoService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:2265
 appDevDebugProjectContainer->getSession_Handler_PdoService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:2304
 appDevDebugProjectContainer->getSession_Storage_NativeService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:2252
 appDevDebugProjectContainer->getSessionService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:1937
 appDevDebugProjectContainer->getSecurity_Csrf_TokenManagerService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:2991
 appDevDebugProjectContainer->getTwigService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/app/cache/dev/appDevDebugProjectContainer.php:347
 appDevDebugProjectContainer->getCacheWarmerService() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:523
 Symfony\Component\HttpKernel\Kernel->initializeContainer() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:133
 Symfony\Component\HttpKernel\Kernel->boot() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:70
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /mnt/tmp/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:123
 Symfony\Component\Console\Application->run() at /mnt/tmp/app/console:29
```

If you would like to see the cache files, here you go - [dev-cache.tar.gz](https://github.com/symfony/symfony/files/204600/dev-cache.tar.gz), archive contains app/cache/dev directory.

## How to reproduce

**TL;DR**

Download pre-configured [symfony27.tar.gz](https://github.com/symfony/symfony/files/204260/symfony27.tar.gz) and [symfony28.tar.gz](https://github.com/symfony/symfony/files/204261/symfony28.tar.gz) with all the steps described below already done. Extract them and inside the project's directory run app/console. If you have Docker installed it can be done as easy as running:

```
sudo rm app/cache/* -rf && docker run -it --rm -v `pwd`:/mnt/tmp -w /mnt/tmp modera/php5-fpm:1.0 app/console
```

---

## Full description how to reproduce the issue

Create **session.yml** file and place it in **app/config** directory.

``` session.yml
parameters:
    pdo.db_options:
        db_table:    foo_sessions_table
        db_id_col:   session_id
        db_data_col: session_value
        db_time_col: session_time
        db_lifetime_col: session_lifetime

services:
    pdo:
        class: PDO
        arguments:
            - "mysql:host=%database_host%;port=%database_port%;dbname=%database_name%"
            - "%database_user%"
            - "%database_password%"
        calls:
            - [setAttribute, [3, 2]] # \PDO::ATTR_ERRMODE, \PDO::ERRMODE_EXCEPTION

    session.handler.pdo:
        class:     Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler
        arguments: ["@pdo", "%pdo.db_options%"]
```

After that update your framework/session configuration in **config.yml** so it would look similar to this:

```
   // ...
    session:
        handler_id: session.handler.pdo
   // ...
```

Make sure that Symfony wouldn't be able to connect to configured in parameters.yml database (the easiest way is to change *database_host* to some random IP).

Make sure that your cache directory is empty by running something like:
```
rm app/cache/* -rf
```

Now you can try to run app/console - in 2.7 version you would get standard list command output whereas 2.8 attempts to connect to a database, though it is not really needed at this moment. It seems that some container compilation or bootstrapping logic has been changed between versions, I haven't yet had time to investigate it.

Any thoughts, guys ?

**Comment by xabbuh** [comment]
Can you share the full stack trace of the error you get?

**Comment by Koc** [comment]
The problem in cache warmers. Cache warmers instanciate twig for compile templates, then `AppVariable` creates and injects csrf token storage which requires session.

```
Z:\home\dev\symfony>php app/console -vvv


  [PDOException]
  SQLSTATE[HY000] [1045] Access denied for user 'root'@'localhost' (using password: YES)


Exception trace:
 () at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:1695
 PDO->__construct() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:1695
 appDevDebugProjectContainer->getPdoService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:2267
 appDevDebugProjectContainer->getSession_Handler_PdoService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:2306
 appDevDebugProjectContainer->getSession_Storage_NativeService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:2254
 appDevDebugProjectContainer->getSessionService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:1939
 appDevDebugProjectContainer->getSecurity_Csrf_TokenManagerService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:2993
 appDevDebugProjectContainer->getTwigService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\app\cache\dev\appDevDebugProjectContainer.php:347
 appDevDebugProjectContainer->getCacheWarmerService() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\DependencyInjection\Container.php:316
 Symfony\Component\DependencyInjection\Container->get() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\HttpKernel\Kernel.php:523
 Symfony\Component\HttpKernel\Kernel->initializeContainer() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\HttpKernel\Kernel.php:133
 Symfony\Component\HttpKernel\Kernel->boot() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Bundle\FrameworkBundle\Console\Application.php:70
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at Z:\home\dev\symfony\vendor\symfony\symfony\src\Symfony\Component\Console\Application.php:123
 Symfony\Component\Console\Application->run() at Z:\home\dev\symfony\app\console:29
```

@sergeil As workaround you can mark `session.handler.pdo` as lazy and install `ocramius/proxy-manager`, see http://symfony.com/doc/current/components/dependency_injection/lazy_services.html for more information

**Comment by Tobion** [comment]
I suggest to let PdoSessionHandler make the connection lazily itself. Instead of passing the PDO instance, you pass the connection DSN. This is also the documented way in http://symfony.com/doc/current/cookbook/doctrine/pdo_session_storage.html by now.

**Comment by sergeil** [comment]
Guys, thank you for the feedback. I have updated the ticket, now it includes the stacktrace as well as archive with app/cache/dev directory, hopefully it helps.

Solution you have offered probably would fix this specific instance of the problem, but the issue seems to be even more serious - although it is more difficult to reproduce, **it seems a similar issue occurs when using custom implementation of UserProviderInterface in security configuration which relies on Doctrine**. I won't post here steps required to reproduce the issue when custom UserProvider is used, root of the problem seems to be of the similar nature. Please ping me, if it is really necessary, I will try then try to extract minimum amount of configuration from our project that is required to reproduce the issue in the context of UserProvider.

Probably in order to better understand the cause of the problem differences in Symfony bootstrapping process can be compared between 2.7 and 2.8 - as far as I understood by reading [UPGRADE-2.8.md](https://github.com/symfony/symfony-standard/blob/2.8/UPGRADE-2.8.md) document some things indeed have changed.

**Comment by jakzal** [comment]
@sergeil I think your initial issue should be solved in the way Tobion described. As a matter of fact you should delay initialization of anything that can't be bootstraped before cache is warmed up. Therefore your doctrine issue might be quite the same problem.

**Comment by xabbuh** [comment]
Closing here as the solution for the initial problem was outlined by @Tobion.

@sergeil If you think that your other issue is really related to Symfony, please fork the Symfony Standard Edition and make the changes that are required to reproduce your issue. Though I guess your issue is probably related to some eager initialisation in a constructor of some of your services.

**Comment by sergeil** [comment]
Thank you guys for your time and help!

In short, solution suggested by @Tobion has indeed solved the problem of app/console crashing on a cold-start. After this problem has been fixed, another one occurred - Symfony has started crashing when building our custom implementation of UserProviderInterface, which was doing $em->getRepository() in constructor method, which eventually resulted in Doctrine trying to establish connection to a database. Our custom implementation of UserProviderInterface had been created in pre-historic times, in hindsight it is already hard to say exactly why we opted to using a custom solution. In order to solve the problem I just refactored our security layer configuration as explained in http://symfony.com/doc/current/cookbook/security/entity_provider.html. So far everything works fine, yay.

**Comment by stof** [comment]
@sergeil when using dependency injection, doing logic in the constructor is not a good idea. And ``$em->getRepository()`` requires lots of work from Doctrine (on first call)

**Comment by sergeil** [comment]
@stof, agree. In our case as it happens at times, especially given that our custom implementation of UserProviderInterface, was created years ago, nobody paid attention because the thing just worked. I have informed other devs in our company about the possible consequences/performance costs incurred by doing heavy-lifting in constructor, so hopefully it won't happen again. But as you said, yes, it is not a good idea in general to perform resource-consuming operations in a class constructor, better approach would be to deffer them to the moment when it is really necessary.

**Comment by Tobion** [comment]
@sergeil see long-standing issue doctrine/DoctrineBundle#351 about that

**Comment by sergeil** [comment]
@Tobion thanks for the reference, seems like a long thread, gonna check it during the weekend :)
